scipy/spatial/distance.py

```diff
@@ -307,6 +311,76 @@ def _validate_weights(w, dtype=np.float64):
     return w
 
 
+def _add_periodic_bounds(fun):
+    _name = 'periodic_bounds'
+    _type = "tuple of ndarray, default: None"
+    _desc = textwrap.dedent(
+        """
+        Tuple containing the lower and upper bounds of periodic boundary conditions,
+        e.g. ``(lb, ub)``.  ``lb`` and ``ub`` must be one-dimensional arrays with the
+        same shape as arguments `u` and `v`, and all of ``ub - lb`` must be positive.
+        Elements of ``ub - lb`` that are non-finite will not be treated as periodic.
+        """
+        ).split('\n')[1:]
+    _periodic_bounds_doc = _docscrape.Parameter(_name, _type, _desc)
+    _periodic_bounds_param = inspect.Parameter(
+        _name, inspect.Parameter.KEYWORD_ONLY, default=None)
+
+    def wrap_to_bounds(x, lb, period):
+        return _lazywhere(np.isfinite(period),
+                          (x, lb, period),
+                          f=lambda x, lb, period: (x - lb) % period + lb,
+                          f2=lambda x, lb, period: x)
+
+    @functools.wraps(fun)
+    def wrapped(*args, periodic_bounds=None, **kwargs):
+        if periodic_bounds is None:
+            return fun(*args, **kwargs)
+        if not isinstance(periodic_bounds, tuple):
+            message = '`periodic_bounds` must be a tuple.'
+            raise ValueError(message)
+        if not len(periodic_bounds) == 2:
+            message = '`periodic_bounds` must contain lower and upper bounds.'
+            raise ValueError(message)
+
+        lb, ub = np.asarray(periodic_bounds[0]), np.asarray(periodic_bounds[1])
+        u = _validate_vector(args[0])
+        v = _validate_vector(args[1])
+
+        if not (ub.shape == lb.shape == u.shape == v.shape):
+            message = 'Lower and upper bounds must have the same shape as `u` and `v`.'
+            raise ValueError(message)
+
+        with np.errstate(invalid='ignore'):
+            period = ub - lb
+
+        if np.any(period <= 0):
+            message = 'Each upper bound must exceed the corresponding lower bound.'
+            raise ValueError(message)
+
+        u = wrap_to_bounds(u, lb, period)
+        v = wrap_to_bounds(v, lb, period)
+        period[~np.isfinite(period)] = 0  # ignore non-finite period
+        uv = u - v
+        uv_sign = np.sign(uv) * period
+        v = u + np.minimum(np.abs(u - v), np.abs(u - v - uv_sign))
+        args = (u, v) + args[2:]
+        return fun(*args, **kwargs)
+
+    doc = _docscrape.FunctionDoc(wrapped)
+    doc['Parameters'].append(_periodic_bounds_doc)
+    doc = str(doc).split("\n", 1)[1]  # remove signature
+    wrapped.__doc__ = str(doc)
+
+    sig = inspect.signature(wrapped)
+    parameters = sig.parameters
+    parameter_list = list(parameters.values())
+    parameter_list.append(_periodic_bounds_param)
+    sig = sig.replace(parameters=parameter_list)
+    wrapped.__signature__ = sig
+    return wrapped
```

Comment:
This construct is pretty hard to read IMO -- I might prefer just adding optional argument(s) to each function directly to be honest, if there really is demand/motivation for it. Common infrastructure could still be abstracted if it really is that common and really handles PBC in a general way for different container/unit cell space geometries.

It looks like a lot like the nan handling and rng handling stuff, which I've always found a bit hard to follow. In `spatial` I'd at least want to be a voice for potentially having a few more lines of code/docs vs. abstraction/overriding that gets a bit hard to follow, but happy to be outvoted as well.

Reply:
Sure. In `spatial`, I would defer to your preferences if this would be accepted otherwise. So before merging, I could eliminate the decorator and copy-paste everything to make the code as verbose as desired, but I'd rather keep it concise before then to avoid duplicate work.

Reply:
I think I was also a bit skeptical that we'd get away with things being that common across the PBC handling scenarios that are possible (i.e., the implementations might fork across metrics depending on how general the PBC handling is). Anyway, I've pinged some folks whom I think are interested in PBC to get some feedback. Maybe it really is that simple but good to ask..

Reply:
I don't think they will be common across all. I applied this only to the metrics for which $d(u, v) = \sum_i f(|u_i - v_i|)$ and $f$ is increasing. It seemed that if there is period $\lambda_i$ in the coordinate $i$, we can take 

```math
  |u_i - v_i| = \min \left( \mod(u_i - v_i, \lambda_i), \quad \lambda_i - \mod(u_i - v_i, \lambda_i) \right)
```

Because $d$ is just a sum of some increasing functions of these terms, it is safe to find the minimum distance by doing the minimization above for each coordinate separately.

Reply:
Yeah, I've just had bad experiences depending on PBC-handling code doing the right thing when I work with different geometries. For example, with cubic-style unit cell, "ok" it usually works, but with splitting over a rhombic dodecahedron I usually can't depend on the scientific Python ecosystem to do the right thing, even in downstream libs that are more specialized.

If you've already thought about all those geometries/scenarios then you're way ahead of me--that's my main point of caution.

Reply:
> If you've already thought about all those geometries

No, I think this makes an assumption about the geometry (which we would document) and hopes that is useful in some applications. This seemed natural enough that it would be useful across domains. I didn't imagine we'd offer options for more exotic geometries in SciPy.